File: php/exercise06.php
<?php

function matrixChainOrder($p, $n) {
    for ($i = 1; $i < $n; $i++)
        $m[$i][$i] = 0;

    for ($L = 2; $L < $n; $L++)
    {
        for ($i = 1; $i <= $n - $L + 1; $i++)
        {
            $j = $i + $L - 1;
            $m[$i][$j] = PHP_INT_MAX;
            for ($k = $i; $k <= $j - 1; $k++)
            {
                // $q = cost/scalar multiplications
                $q = @$m[$i][$k] + @$m[$k + 1][$j] + @$p[$i - 1] * @$p[$k] * @$p[$j];
                if ($q < $m[$i][$j])
                    $m[$i][$j] = $q;
            }
        }
    }
    return $m[1][$n-1];
}

$arr = array(54,66,39,46,12,87,61,24,36,28,13,4,73,84,25,62,67);
$size = sizeof($arr);

echo "Minimum number of multiplications is ".matrixChainOrder($arr, $size);

?>

File: python/exercise06.py
# A naive recursive implementation that 
# simply follows the above optimal 
# substructure property 
import sys 

# Matrix A[i] has dimension p[i-1] x p[i] 
# for i = 1..n 
def MatrixChainOrder(p, i, j): 

	if i == j: 
		return 0

	_min = sys.maxsize 
	
	# place parenthesis at different places 
	# between first and last matrix, 
	# recursively calculate count of 
	# multiplications for each parenthesis 
	# placement and return the minimum count 
	for k in range(i, j): 
	
		count = (MatrixChainOrder(p, i, k) 
			+ MatrixChainOrder(p, k+1, j) 
				+ p[i-1] * p[k] * p[j]) 

		if count < _min: 
			_min = count; 
	

	# Return minimum count 
	return _min; 


# Driver program to test above function 
arr = [54,66,39,46,12,87,61,24,36,28,13,4,73,84,25,62,67]; 
n = len(arr); 

print("Minimum number of multiplications is ", 
				MatrixChainOrder(arr, 1, n-1)); 

# This code is contributed by Aryan Garg 
